File: Data Structures & Algorithms/binary-search/submission-2.java
class Solution {
    public int search(int[] nums, int target) {
        int L = 0;
        int R = nums.length-1;
        int mid = (L+R)/2;
        System.out.println(mid);
        while(L<=R){
            mid = (L+R)/2;
            System.out.println("------");
            System.out.println(L);
            System.out.println(mid);
            System.out.println(R);
            System.out.println("------");
            if (nums[mid] == target){
                return mid;
            }else if(  target < nums[mid]){
                R = mid - 1;
            }else{
                L = mid + 1;
            }
        }
        return -1;
    }
}
/*

L = 0, R = n-1
while(L<R):
    mid = (L + R)/2
    if target = nums[mid]:
        return mid
    else if target <nums[mid]:
        R = mid-1
    else:
        L = mid+1
*/

File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        '''
        create a min heap of size k 
        insert the freq and element as a tuple to the heap O(n) O(n.logk)
        if size is greater than k
        remove the current minimum 

        collect all elements from the min heap and return O(n)


        '''
        freq = Counter(nums)
        heap = []
        for i in freq.items():
            heapq.heappush(heap,(i[1],i[0]))
            if(len(heap)>k):
                heapq.heappop(heap)
        return [x[1] for x in heap]